src/modm/ui/color/gray.hpp

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2022, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <algorithm>
+#include <concepts>
+#include <functional>
+
+#include <modm/math/saturation/saturated.hpp>
+#include <modm/math/proportional_unsigned.hpp>
+
+#include <modm/math/utils/arithmetic_traits.hpp>
+#include <modm/math/utils/misc.hpp>
+
+#include "concepts.hpp"
+
+namespace modm::color {
+/**
+ * @brief 		Unsigned integer with arbitrary number of digits, proportional conversion
+ * 				and saturating arithemtics.
+ *
+ * @tparam D 	Number of Digits
+ * @tparam P	When forming a Buffer with this color, this would be the internal storage.
+ * 				Should be MCUs fastest unsigned int -> just keep the default.
+ * 				Some Displays require a specific type therefore it's a template argument.
+ *
+ * @author		Thomas Sommer
+ * @ingroup		modm_ui_color
+ */
+template <int D>
+requires (D > 0)
+class GrayD : public modm::ProportionalUnsigned<D>
+{
+	using ProportionalUnsigned<D>::ProportionalUnsigned;
+
+public:
+	using T = ProportionalUnsigned<D>::T;
+
+	using modm::ProportionalUnsigned<D>::operator=;
+
+  	template<ColorRgb C>
+	constexpr GrayD(const C& rgb)
+		: ProportionalUnsigned<D>((
+			2125 * (modm::WideType<T>)(GrayD(rgb.red())) +
+			7154 * (modm::WideType<T>)(GrayD(rgb.green())) +
+			0721 * (modm::WideType<T>)(GrayD(rgb.blue()))
+			) / 10000
```

Comment:
Could we change the constants in a way that the number we divide by is a power of 2? Then the divide is a left shift. That would be more efficient, especially on Cortex-M0(+) or AVR.

Reply:
I've rounded down:
2125 -> 2124
721 -> 720

Reply:
I think I got that from [the Wikipedia Grayscale article](https://en.wikipedia.org/wiki/Grayscale#Converting_colour_to_grayscale). The formula seems to be slightly different though: 

![](https://wikimedia.org/api/rest_v1/media/math/render/svg/f84d67895d0594a852efb4a5ac421bf45b7ed7a8)

A possible adaptation for a bit shift value could be.

- Scaled by 2^13/10000: `(1742 * R + 5859 * G + 591 * B) >> 13`
- Scaled by 2^14/10000: `(3483 * R + 11718 * G + 1183 * B) >> 14`

Reply:
Calculation takes place in 32 bit anyway.
'Scaled by 2' seems to be the right pick.

Reply:
I've dump copied and tested your _Scaled by 2^13/10000_ @salkinium . Works like expected, thanks for tinkering this out 👍‍

-> If needed, let's optimize this later in a real application or once the Buffers and Displays got their PR-updates. Will be much easier to confirm the conversion quality if one can generate and inspect interference patterns and such on a screen.